### PR TITLE
View app pages based on role

### PR DESCRIPTION
**Pull request by john-tco**
- GAP-1942
- related pr: https://github.com/cabinetoffice/gap-user-service/pull/76/

### CODE REVIEW

File: packages/admin/src/pages/index.page.tsx [hunk 1]
```diff
@@ -50,4 +44,23 @@ const index = () => {
   return <div>index</div>;
 };
 
+const handleError = (error: any) => {
+  console.log('Failed to verify token', error);
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/index.page.tsx:48]: This can be console.error

File: packages/admin/src/pages/index.page.tsx [hunk 2]
```diff
@@ -50,4 +44,23 @@ const index = () => {
...
+
+  if (error.response?.data?.error?.message === 'User is not an admin') {
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/index.page.tsx:50]: Is there any other way we can assert that the user is not an admin? relying the error message is not an ideal approach.

Comment (john-tco) [reply packages/admin/src/pages/index.page.tsx]: Sure, using the status now

File: packages/admin/src/pages/index.page.tsx [hunk 3]
```diff
@@ -50,4 +44,22 @@ const index = () => {
   return <div>index</div>;
 };
 
+const handleError = (error: any) => {
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/index.page.tsx:47]: is error an AxiosError?

File: packages/admin/src/pages/index.page.tsx [hunk 4]
```diff
@@ -50,4 +44,22 @@ const index = () => {
...
+  console.error('Failed to verify token', error);
+  if (error.response?.status === 401) {
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/index.page.tsx:49]: `error?.response?.status ` will ensure we don't get a type error if error is null or undefined

File: packages/admin/src/pages/index.page.tsx [hunk 5]
```diff
@@ -14,14 +16,8 @@ export const getServerSideProps: GetServerSideProps = async ({
...
     response = await authenticateUser(cookieValue);
-  } catch (error) {
-    console.error('Failed to verify token', error);
-    return {
-      redirect: {
-        destination: getLoginUrl(),
-        permanent: false,
-      },
-    };
+  } catch (error: unknown) {
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/index.page.tsx:19]: this be error: AxiosError too or else we will get a type error down below

Comment (john-tco) [reply packages/admin/src/pages/index.page.tsx:19]: You cannot actually add types to arguments in catch
>Catch clause variable type annotation must be 'any' or 'unknown' if specified.ts(1196)

File: packages/admin/src/pages/super-admin-dashboard/manage-departments/index.page.tsx [hunk 6]
```diff
@@ -13,13 +13,19 @@ import { GetServerSidePropsContext } from 'next';
...
   const fetchPageData = async (jwt: string) => ({
     departments: await getAllDepartments(jwt),
     userId: context.query?.userId || '',
   });
-  return { props: await fetchPageData(getUserTokenFromCookies(context.req)) };
+  try {
+    return { props: await fetchPageData(getUserTokenFromCookies(context.req)) };
+  } catch (err) {
+    return handleAxiosError(err as AxiosError);
+  }
```
Comment (dominicwest) [inline packages/admin/src/pages/super-admin-dashboard/manage-departments/index.page.tsx:28]: Could abstract this logic into the new function?

Would need to accept a function as a param, generically type it, and return either the functions response, or if there's an error the logic you already have there

File: packages/admin/src/utils/handleAxiosError.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      },
+    };
+  }
+  return {
+    redirect: {
+      destination: getLoginUrl(),
+      permanent: false,
+    },
+  };
+};
```
Comment (jgunnCO) [inline packages/admin/src/utils/handleAxiosError.ts:22]: I think the name here is a bit misleading as this is actually more of a general error handler to run while fetching data, it doesn't just handle axios errors and the error passed doesn't really need to be one. I would just type it to accept an `unknown` argument and give it more general name, and then check the type in the function like this:

```suggestion
const handleRequestError = (error: unknown) => {
  console.error('Failed to verify token', error);
  if (error instanceof AxiosError) {
    const unauthorized =
      error.response?.status === 401 || error.response?.status === 403;
    if (unauthorized) {
      return {
        redirect: {
          destination: '/404',
          permanent: false,
        },
      };
    }
  }
  return {
    redirect: {
      destination: getLoginUrl(),
      permanent: false,
    },
  };
};
```

File: packages/admin/src/pages/super-admin-dashboard/user/[id]/index.page.tsx [hunk 8]
```diff
@@ -7,15 +7,24 @@ import { User } from '../../types';
...
+  if ('redirect' in pageData) {
+    return pageData;
+  }
+
   return {
     props: {
-      user: await getUserById(params?.id as string, userToken),
+      user: pageData,
     },
   };
```
Comment (dominicwest) [inline packages/admin/src/pages/super-admin-dashboard/user/[id]/index.page.tsx:29]:
```
const getPageData = () => getUserById(params?.id as string, userToken);
return await fetchDataOrGetRedirect(getPageData);

File: packages/admin/src/utils/fetchDataOrGetRedirect.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+async function fetchDataOrGetRedirect<T>(
+  fetchData: (...args: unknown[]) => Promise<T> | { redirect: object }
+): Promise<ReturnType<typeof fetchData>> {
+  try {
+    return await fetchData();
+  } catch (error: unknown) {
+    return getRedirect(error);
+  }
+}
```
Comment (dominicwest) [inline packages/admin/src/utils/fetchDataOrGetRedirect.ts:12]:
```suggestion
async function fetchDataOrGetRedirect<T>(
  fetchData: (...args: unknown[]) => Promise<T>
): Promise<{ props: T } | { redirect: Redirect }> {
  try {
    return { props: await fetchData() };
  } catch (error: unknown) {
    return getRedirect(error);
  }
}
```

**Review by dominicwest — CHANGES_REQUESTED** [review]

**Review by dominicwest — APPROVED** [review]

**Pull request merged**